src/high-level/constructors.lisp

```diff
@@ -0,0 +1,181 @@
+;;;; constructors.lisp
```

Comment:
Just a suggestion (again, this PR covers a lot of ground, so I don't view this as a requirement or anything), it would be nifty to have an additional constructor which allows for one to generate a tensor by specifying index variables, corresponding dimensions, and a generating expression for the actual tensor entries. For example, I whipped together this macro (I think there are similar things for other multidimensional array libraries):

```
(defmacro generate-tensor (type index-exprs &body body-expr)
  "Generate an expression which evaluates to a MAGICL TENSOR of the given TYPE.

Here, INDEX-EXPRS is an expression of the form
   ((i0 d0) (i1 d1) ... (iN dN)).
Here D0, D1, ..., DN denote the shape of the resulting tensor. The variables
I0, I1, ... IN will be bound in BODY-EXPR, and with the value of BODY-EXPR
determining the value of the (I0, I1, ..., IN) entry of the tensor. "
  (alexandria:with-gensyms (shape tensor)
    (let ((vars (mapcar #'first index-exprs)))
      (labels ((generate-loops (i)
                 (if (= i (length vars))
                     `(setf (magicl:tref ,tensor ,@vars)
                            (coerce (progn ,@body-expr) ,type))
                     `(dotimes (,(elt vars i)
                                (elt ,shape ,i))
                        ,(generate-loops (1+ i))))))
        `(let* ((,shape (list ,@(mapcar #'second index-exprs)))
                (,tensor (magicl:const 0.0 ,shape :type ,type)))
           ,(generate-loops 0)
           ,tensor)))))
```

so that we can do something like

```
> (generate-tensor 'single-float ((i 3) (j 2))
  (+ i j))

#<MAGICL:MATRIX/SINGLE-FLOAT (3x2):
   0.000     1.000
   1.000     2.000
   2.000     3.000>

```

Reply:
#71